### ISSUE TITLE
Add ability to assign (non racked) devices to rack groups (aka rooms)

### ISSUE THREAD

**Issue by dreng** [netbox-community/netbox#4971]
### Environment
* Python version:  3.7.3
* NetBox version:  2.8.9

### Proposed Functionality
If I understand #1779 correctly you are not planning to implement rooms (which is extremely sad imo) and it is suggested to abuse rack groups to act as a room. If this is the case one should have the ability to assign any (non racked) device to such a group. Currently it is only possible to add racks to rack groups. Unfortunately the real world does not only consist of racks.

### Use Case
Dozens. E.g. UPS or emergency power generator in a different fire section, non racked and seperated from all other rooms.

For a datacenter admin (and every other admin as well) it should be clear at first sight where a device is located. I do not just need to have a list of all devices but a list of where the devices can be found. Imagine a broken device that needs physical access to be repaired. Where should one start to search if there are plenty of rooms equipped with non racked devices? Sure, you can write a room number into a comment field but that totally breaks the intention of connecting data with parent/child relations. There wouldn't be any chance to use that data in a reasonable way.

**Comment by mskalecki** [comment]
Rack Groups can't have devices assigned directly to them.  However, Racks can have 0U devices assigned to them (just leave the Rack Face and Position fields blank).

If you don't want unracked devices associated with "real" racks.  Add additional imaginary racks that describe the location of the unracked devices (like "WALL-1"), and assign the unracked devices to those "racks".  This works great.

**Comment by dreng** [comment]
@mskalecki 
Thank you for your contribution. Generally speaking, it is a very bad idea to abuse database fields or functions for things they were not invented for. In your example you can see directly, where such ideas end. Just click "Organisation -> Elevations" and find a bunch of physically not existing Racks in a graphical overview. Things will definately become worse with future developments.

Another (very bad) idea would be to allow 0U-Racks, which is currently not possible. Funny enough it is possible to choose a value of "0"  at "Height (U)" which ends in an error message "Ensure that value is greater than or equal to 1."

Don't get me wrong as I do not intend to be harsh. I just don't get the point why there is no room concept as it is essential imo. Instead the official way is to break a great design with a poor workaround. And if that workaround does not work the solution is to use an even worse workaround. Furthermore it should be very easy to implement another layer between sites and racks. Racks should belong to a site XOR room. If one thinks that this is out of scope of a DCIM, have a look at [this whitepaper](https://it-resource.schneider-electric.com/small-medium-data-centers/wp-104-classification-of-data-center-infrastructure-management-dcim-tools) from Schneider Electric.

**Comment by mskalecki** [comment]
@dreng 
I agree with your premise of not abusing database fields, but my reasoning comes to a completely different conclusion.

In Netbox, "Racks" are a model of a physical location where devices physically reside.  Typically, this will be a traditional network or server rack (hence the name), but a wall or section of flooring where devices reside is logically identical to traditional rack, except that the concept of an elevation or strict size limit may not make sense (Netbox already accomodates this by providing a section for "nonracked" devices on the elevation page).  While one could separate out racks with elevations and nonracked areas for devices into distinct models, there is no real gain here, and you over-complicated the case where a tradional rack has associated device that don't fit in rack slots (such as a vertical PDU).

If you mentally imagine the Netbox "Rack" model as "Device Group", and then create Rack Roles for each type of device group you want to model ("network rack", "server rack", "wall", "cabinet", etc).  The logic is sound.

Meanwhile, Rack Groups, are simply a groupings of Racks and other Rack Groups.  Assigning devices directly to a Rack Group would be a pretty big logical change that only removes information.

**Comment by proudbro** [comment]
@mskalecki , why don't you create rooms as sites rather than rack groups? We've made global regions as Regions and buildings as Nested Regions. So you can attach devices (UPS, CRAC and so on) to sites and use rack groups to represent the rows

**Comment by dreng** [comment]
Ok, I got #1779 completely wrong then. Maybe there's a language barrier to blame.

So, you suggest to use racks with non racked devices for grouping purposes and let them act as a room. Although that doesn't seem to be ideal nor logical, i could live with it. Problem is, there seems to be no way of not showing such "rooms" as empty racks at the elevations overview, or am I missing something? And that might be not the only side effect. A rack (device group, room) still has a width of 19" and a height of x rack units. That's what I meant when i wrote about abusing database fields. There will always be one day such workarounds break your neck. As a compromise, maybe a boolean rack attribute llike "this is a virtual rack" could work around the issue? You could show all non-virtual racks in the elevations overview then.

I don't completely agree to your logic of racks, device groups and rack roles. I understand that you can imagine racks as logical groups. What I'd really like to see some day (maybe as a plugin) is a ground view with all your IT stuff (as seen in some commercial DCIM solutions). How should such a plugin know if a rack should be drawn as a rack, a room or maybe not at all? @proudbro found another solution to work around that problem which makes things even more complicated. There's just no best practice or an intuitive way of doing it. Furthermore I don't get the point of rack groups then. If it's just a matter of sound, it should be ok to nest one or more racks in a rack (aka room or rack group). No need for rack groups anymore here. Just imagine a rack as a rack group.
What I'm trying to say: Of course you can imagine everything as everything else, but that is not what a software makes a good solution. It's not just a matter of sound but a matter of logic, usability and intuitive operation.

Long story short: We're just tinkering around at the moment and I have to discuss with my team if we've got a show stopper here. As asked before, would it be a solution to add the ability to make a rack "virtual"?

**Comment by minitriga** [comment]
We use netbox to manage our "Lab", which are racks of gear used to test customer networks. However, we have an issue whereby we have more gear than racks available due to customers wanting to test their own hardware in our facility with Ixia etc so we needed to be able to assign a device to our storage room. To do this we had to create a rack group "Storage Room" and then assign racks to said group in which I can put spares and unused devices. Bit of a pain to be honest. I have created a few plugins so might tackle this myself.

**Comment by jeremystretch** [comment]
In the interest of summarizing the discussion above, I've sketched the diagram below. The solid lines indicate existing relationships, and the dashed line represents my understanding of the proposed new relationship.

![device_hierarchy](https://user-images.githubusercontent.com/13487278/90677888-fd568b80-e22b-11ea-8673-846970fb66fb.png)

The assignment of a device to a site is required; association with a particular rack (with or without a specific face/position) is optional. While I don't see any obvious blockers to accommodating the additional relationship, we would need to take care to ensure proper validation is implemented (e.g. to make the appropriate updates when a RackGroup is assigned to a new Site). However we should be able to handle that in a manner similar to the current validation for racks.

My assumption is that any device assigned to a rack would be automatically associated with that rack's group (as opposed to making the Rack and RackGroup assignments mutually exclusive).

**Comment by jeremystretch** [comment]
Moving this to `needs milestone` with the understanding that the goal is to add a relation from Device to RackGroup.

**Comment by candlerb** [comment]
@jeremystretch wrote:

> In the interest of summarizing the discussion above, I've sketched the diagram below. The solid lines indicate existing relationships, and the dashed line represents my understanding of the proposed new relationship.

I think that diagram is incomplete, as:
1. There is also a direct link from Rack to Site, which is not shown
2. It doesn't show the recursive parent relationship of RackGroup (that would just be a loopback arrow)

> While I don't see any obvious blockers to accommodating the additional relationship, we would need to take care to ensure proper validation is implemented (e.g. to make the appropriate updates when a RackGroup is assigned to a new Site).

I think those issues already exist, namely:
- if a RackGroup is assigned to a Site, then this may or may not be consistent with the Racks contained within it
- if a RackGroup is assigned to a Site, then this may or may not be consistent with its child and/or parent RackGroups 

Checking the existing validation rules, I find the following:
- `Rack.clean` checks that if you change the Site of a Rack, it's consistent with the Site of its immediate parent RackGroup (if any).
- `RackGroup.clean` checks that if you change the Site of a RackGroup, it's consistent with the Site of its immediate parent RackGroup (if any).

However I don't see any checks in the reverse direction: e.g. if you change the Site of a RackGroup it doesn't check that its child and descendant RackGroups and Racks are in the same Site [^1].  Indeed, testing (v2.9.8) shows this check is missing:

* Create two sites, A and B
* Create rack X in site A
* Create rackgroup G in site A
* Assign rack X to rackgroup G
* Change site of rackgroup G to B

This change is accepted: rack X now belongs to site A but a rackgroup in site B. [^2]

Therefore: I don't think being able to assign a device directly to a Rackgroup adds any more requirements to validate the Rackgroup tree, beyond the (missing) functionality already required.

There is also validation around the Device's Site versus the Device's Rack's Site.  Again this is missing when you change the Site of a Rack, which can be demonstrated as follows:

* Create two sites, A and B
* Create rack X in site A
* Create devices D1, D2, D3... in rack X
* Change rack X's site to B

Now those devices exist in site A, but in a rack which is in site B.

So in summary: if a `device->rackgroup` link is added, then `device->rackgroup->site` will also need to be validated, but this is no more than the already-required (but missing) `device->rack->site` validation. [^3]


-------

[^1] Also, without a cascading downward change it becomes very difficult to change the assigned site of a tree of rackgroups. I think you'd have to replicate the entire rackgroup tree in a new site, reassign the racks to new site and rackgroup simultaneously, and then delete the original rackgroup tree.  However this will be required very rarely, if ever.

[^2] From this point onwards, it also becomes impossible to make an unrelated change to rack X: e.g. if you try to add a comment on rack X, it's rejected with "Rack group must be from the same site".

[^3] There is a possibly stronger argument here for a cascading change.  Otherwise, moving a rack from site X to Y would require creating a new rack in Y, changing all the devices to the new rack, and then deleting the old rack.  However again this really shouldn't happen much in practice.

**Comment by candlerb** [comment]
I support linking Device directly to RackGroup (representing "room" or "area" within Site).

There is a precedent: PowerPanel is linked to Site and optionally RackGroup, but not to Rack.

**Comment by silvanet123** [comment]
I agree about having some setup to represent a room or area.  I understand that NetBox is primarily for DCIM; however, having rooms or areas where devices and data drops could go into would be EXTREMELY valuable.   I am migrating from a similar tool I have used for years, Netdot (https://github.com/cvicente/Netdot), to NetBox.  Netdot has similar features; however, I like the direction and activity of NetBox.  The feature of cable plants in Netdot what initially appealing for me, particularly to rooms.

**Comment by jeremystretch** [comment]
Blocked by #5895 (renaming RackGroup)
